network/transport/v2/retry.go

```diff
@@ -0,0 +1,195 @@
+/*
+ * Nuts node
+ * Copyright (C) 2021 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package v2
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+	"github.com/nuts-foundation/nuts-node/network/log"
+	"go.etcd.io/bbolt"
+)
+
+// Retriable defines methods for a persistent retry mechanism
+type Retriable interface {
+	// Add a job that needs to be retried. It'll be passed to the channel immediately and at set intervals.
+	// Returns nil if job already exists.
+	// An error is returned if there's a problem with the underlying storage.
+	Add(hash hash.SHA256Hash) error
+	// Remove a job, it'll no longer be retried.
+	// An error is returned if there's a problem with the underlying storage.
+	Remove(hash hash.SHA256Hash) error
+	// Configure opens the DB connection for persistent job storage and loads existing jobs from disk
+	Configure() error
+	// Start retrying new and existing jobs
+	Start() error
+	// Close cancels all jobs and closes the DB
+	Close() error
+}
+
+type jobCallBack func(hash hash.SHA256Hash)
+
+// NewPayloadRetrier returns a Retriable for payload fetches.
+// The payload hashes as []byte should be added as job.
+func NewPayloadRetrier(config Config, callback jobCallBack) Retriable {
+	return &payloadRetrier{
+		config:   config,
+		callback: callback,
+	}
+}
+
+type payloadRetrier struct {
+	config   Config
+	db       *bbolt.DB
+	callback jobCallBack
+	ctx      context.Context
+	cancel   context.CancelFunc
+}
+
+func (p *payloadRetrier) Configure() error {
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+
+	dbFile := path.Join(p.config.Datadir, "network", "payload_jobs.db")
+	if err := os.MkdirAll(filepath.Dir(dbFile), os.ModePerm); err != nil {
+		return fmt.Errorf("unable to setup database: %w", err)
+	}
+
+	if p.config.PayloadRetryDelay == 0 {
+		p.config.PayloadRetryDelay = 5 * time.Second
+	}
+	var err error
+	p.db, err = bbolt.Open(dbFile, 0600, bbolt.DefaultOptions)
```

Comment:
wouldn't it be more flexible to let the retry db be passed by the caller, instead of the retrier always making a database by itself?

Reasoning: maybe the payload_jobs bucket should be part of the DAG BBolt database. Because the DAG TXs and shouldn't be backup'd apart from each other (otherwise you might be missing payloads after copying/restoring), the same goes for the payload jobs: if you fail to restore/copy those, there might be private TXs that are never retried?

Reply:
One of the next big things to do is streamline the storage requirements for all engines/components. This relates to DB backup requirements. This DB would then be refactored as well, to get this piece of functionality forward, I chose for a simple local DB.